File: regions/business/taskmanagement/client/flutter/project_master/lib/models/project_type.dart
/// プロジェクトタイプモデル定義
/// タスク管理プロジェクトマスタのプロジェクトタイプを表す
library;

/// プロジェクトタイプモデル
/// プロジェクトの種類を表す（例：ソフトウェア開発、インフラ構築など）
class ProjectType {
  final String id;
  final String code;
  final String displayName;
  final String? description;
  final String? defaultWorkflow;
  final bool isActive;
  final int sortOrder;
  final String createdBy;
  final DateTime createdAt;
  final DateTime updatedAt;

  const ProjectType({
    required this.id,
    required this.code,
    required this.displayName,
    this.description,
    this.defaultWorkflow,
    required this.isActive,
    required this.sortOrder,
    required this.createdBy,
    required this.createdAt,
    required this.updatedAt,
  });

  /// JSONマップからProjectTypeインスタンスを生成する
  factory ProjectType.fromJson(Map<String, dynamic> json) {
    return ProjectType(
      id: json['id'] as String,
      code: json['code'] as String,
      displayName: json['display_name'] as String,
      description: json['description'] as String?,
      defaultWorkflow: json['default_workflow'] as String?,
      isActive: json['is_active'] as bool,
      sortOrder: json['sort_order'] as int,
      createdBy: json['created_by'] as String,
      createdAt: DateTime.parse(json['created_at'] as String),
      updatedAt: DateTime.parse(json['updated_at'] as String),
    );
  }

  /// ProjectTypeインスタンスをJSONマップに変換する
  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'code': code,
      'display_name': displayName,
      'description': description,
      'default_workflow': defaultWorkflow,
      'is_active': isActive,
      'sort_order': sortOrder,
      'created_by': createdBy,
      'created_at': createdAt.toIso8601String(),
      'updated_at': updatedAt.toIso8601String(),
    };
  }
}

/// プロジェクトタイプ作成時の入力モデル
/// 新規プロジェクトタイプ作成に必要なフィールドのみを定義する
class CreateProjectTypeInput {
  final String code;
  final String displayName;
  final String? description;
  final String? defaultWorkflow;
  final bool isActive;
  final int sortOrder;

  const CreateProjectTypeInput({
    required this.code,
    required this.displayName,
    this.description,
    this.defaultWorkflow,
    this.isActive = true,
    this.sortOrder = 0,
  });

  /// 入力データをAPIリクエスト用のJSONマップに変換する
  Map<String, dynamic> toJson() {
    return {
      'code': code,
      'display_name': displayName,
      'description': description,
      'default_workflow': defaultWorkflow,
      'is_active': isActive,
      'sort_order': sortOrder,
    };
  }
}

/// プロジェクトタイプ更新時の入力モデル
/// 更新可能なフィールドのみを定義する
class UpdateProjectTypeInput {
  final String? displayName;
  final String? description;
  final String? defaultWorkflow;
  final bool? isActive;
  final int? sortOrder;

  const UpdateProjectTypeInput({
    this.displayName,
    this.description,
    this.defaultWorkflow,
    this.isActive,
    this.sortOrder,
  });

  /// 更新データをAPIリクエスト用のJSONマップに変換する
  /// nullのフィールドは除外する
  Map<String, dynamic> toJson() {
    final json = <String, dynamic>{};
    if (displayName != null) json['display_name'] = displayName;
    if (description != null) json['description'] = description;
    if (defaultWorkflow != null) json['default_workflow'] = defaultWorkflow;
    if (isActive != null) json['is_active'] = isActive;
    if (sortOrder != null) json['sort_order'] = sortOrder;
    return json;
  }
}

File: regions/business/taskmanagement/server/rust/project-master/src/domain/entity/project_type.rs
// プロジェクトタイプエンティティ。
// ソフトウェア開発・マーケティング等のテンプレート定義を表現する。
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use uuid::Uuid;

/// プロジェクトタイプ（会計の MasterCategory に相当）
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ProjectType {
    pub id: Uuid,
    pub code: String,
    pub display_name: String,
    pub description: Option<String>,
    pub default_workflow: Option<serde_json::Value>,
    pub is_active: bool,
    pub sort_order: i32,
    pub created_by: String,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

/// プロジェクトタイプ作成 DTO
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CreateProjectType {
    pub code: String,
    pub display_name: String,
    pub description: Option<String>,
    pub default_workflow: Option<serde_json::Value>,
    pub is_active: Option<bool>,
    pub sort_order: Option<i32>,
}

/// プロジェクトタイプ更新 DTO
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct UpdateProjectType {
    pub display_name: Option<String>,
    pub description: Option<String>,
    pub default_workflow: Option<serde_json::Value>,
    pub is_active: Option<bool>,
    pub sort_order: Option<i32>,
}

/// プロジェクトタイプ一覧フィルタ
#[derive(Debug, Clone, Default)]
pub struct ProjectTypeFilter {
    pub active_only: bool,
    pub limit: Option<i64>,
    pub offset: Option<i64>,
}

#[cfg(test)]
#[allow(clippy::unwrap_used)]
mod tests {
    use super::*;

    // ProjectTypeエンティティが正しくフィールドを保持することを確認する
    // 前提: 有効なUUIDと文字列を渡す
    // 期待: 各フィールドが正確に格納されている
    #[test]
    fn test_project_type_fields() {
        let id = Uuid::new_v4();
        let now = chrono::Utc::now();
        let pt = ProjectType {
            id,
            code: "SOFTWARE".to_string(),
            display_name: "ソフトウェア開発".to_string(),
            description: Some("説明文".to_string()),
            default_workflow: None,
            is_active: true,
            sort_order: 1,
            created_by: "admin".to_string(),
            created_at: now,
            updated_at: now,
        };
        assert_eq!(pt.id, id);
        assert_eq!(pt.code, "SOFTWARE");
        assert_eq!(pt.display_name, "ソフトウェア開発");
        assert!(pt.description.is_some());
        assert!(pt.is_active);
        assert_eq!(pt.sort_order, 1);
    }

    // CreateProjectType DTOが省略可能フィールドのデフォルト値（None）を取ることを確認する
    // 前提: is_active と sort_order を指定しない
    // 期待: 省略可能フィールドが None となる
    #[test]
    fn test_create_project_type_optional_fields() {
        let input = CreateProjectType {
            code: "MARKETING".to_string(),
            display_name: "マーケティング".to_string(),
            description: None,
            default_workflow: None,
            is_active: None,
            sort_order: None,
        };
        assert_eq!(input.code, "MARKETING");
        assert!(input.is_active.is_none());
        assert!(input.sort_order.is_none());
    }

    // UpdateProjectType DTOが全フィールド省略可能であることを確認する
    // 前提: 全フィールドを指定しない
    // 期待: 全て None となる
    #[test]
    fn test_update_project_type_all_none() {
        let input = UpdateProjectType {
            display_name: None,
            description: None,
            default_workflow: None,
            is_active: None,
            sort_order: None,
        };
        assert!(input.display_name.is_none());
        assert!(input.is_active.is_none());
    }

    // ProjectTypeFilterのデフォルト値が正しいことを確認する
    // 前提: Default トレイトを使用する
    // 期待: active_only=false, limit/offset が None
    #[test]
    fn test_project_type_filter_default() {
        let filter = ProjectTypeFilter::default();
        assert!(!filter.active_only);
        assert!(filter.limit.is_none());
        assert!(filter.offset.is_none());
    }

    // ProjectTypeFilterにactive_only=trueを設定できることを確認する
    // 前提: active_onlyにtrueを指定する
    // 期待: フィルタがアクティブのみに限定される
    #[test]
    fn test_project_type_filter_active_only() {
        let filter = ProjectTypeFilter {
            active_only: true,
            limit: Some(10),
            offset: Some(0),
        };
        assert!(filter.active_only);
        assert_eq!(filter.limit, Some(10));
        assert_eq!(filter.offset, Some(0));
    }
}
